tensorflow-framework/src/main/java/org/tensorflow/framework/metrics/impl/MetricsHelper.java

```diff
@@ -0,0 +1,236 @@
+            tf.constant("weights.shape="),
+            weightsShape,
+            tf.constant("values.shape="),
+            valuesShape,
+            tf.constant("is_scalar="),
+            is_scalar);
+
+    Operand<TBool> isValidShape =
+        tf.select(
+            is_scalar,
+            is_scalar,
+            hasValidNonscalarShape(tf, weightsRank, weightsShape, valuesRank, valuesShape));
+
+    return tf.withSubScope("broadcastWeights-dynamic").assertThat(isValidShape, data);
+  }
+
+  /**
+   * Gets an operand that tests if the shapes have the same rank and valid dimensions.
+   *
+   * @param tf the TensorFlow Ops
+   * @param weightsRank the operand for the rank of the sample weights
+   * @param weightsShape the operand for the shape of the sample weights
+   * @param valuesRank the operand for the rank of the sample weights
+   * @param valuesShape the operand for the shape of the sample weights
+   * @param <T> the data type for the operands
+   * @return a boolean operand to determine if the Shape is scalar or not.
+   */
+  private static <T extends TNumber> Operand<TBool> hasValidNonscalarShape(
+      Ops tf,
+      Operand<T> weightsRank,
+      Operand<T> weightsShape,
+      Operand<T> valuesRank,
+      Operand<T> valuesShape) {
+    tf = tf.withSubScope("has_valid_nonscalar_shape");
+    Operand<TBool> isSameRank = tf.math.equal(valuesRank, weightsRank);
+    return tf.select(isSameRank, hasValidDims(tf, weightsShape, valuesShape), isSameRank);
+  }
+
+  /**
+   * Gets an operand that tests if the shapes have valid dimensions or not.
+   *
+   * @param tf the TensorFlow Ops
+   * @param weightsShape the operand for the shape of the sample weights
+   * @param valuesShape the operand for the shape of the sample weights
+   * @param <T> the data type for the operands
+   * @return a boolean operand to determine if the shapes have valid dimensions or not.
+   */
+  private static <T extends TNumber> Operand<TBool> hasValidDims(
+      Ops tf, Operand<T> weightsShape, Operand<T> valuesShape) {
+    tf = tf.withSubScope("has_invalid_dims");
+    Operand<T> diff = tf.reduceSum(tf.math.sub(weightsShape, valuesShape), tf.constant(0));
+    return tf.math.equal(cast(tf, tf.constant(0), diff.asOutput().type()), diff);
+  }
+
+  // alias for mean
+
+  /**
+   * Calculate the mean of the operand, along all axes and <code>keepDims</code> is <code>false
+   * </code>
+   *
+   * @param tf the TensorFlow Ops
+   * @param x the Operand used to calculate the mean
+   * @param <T> the type of the Operand.
+   * @return the mean of the operand
+   */
+  public static <T extends TType> Operand<T> mean(Ops tf, Operand<T> x) {
+    return mean(tf, x, null, false);
+  }
+
+  /**
+   * Calculate the mean of the operand, alongside the specified axis with <code>keepDims</code> is
+   * <code>false</code>
+   *
+   * @param tf the TensorFlow Ops
+   * @param x the Operand used to calculate the mean
+   * @param axis Axes to compute the mean.
```

Comment:
axis or axes? (We currently say both.)

Reply:
It should be `axes`, but `tf.math.mean` says `axis`, but the description says `dims`, plural.
I change it to `axes`.

Reply:
👍 done